src/config/mod.rs

```diff
@@ -1283,4 +1320,373 @@ make_backup = false
+            let toml = "required_version=\"!=1.0.0\"";
+            let config = Config::from_toml(toml, Path::new("./rustfmt.toml")).unwrap();
+
+            assert!(!config.version_meets_requirement());
+        }
+    }
+
+    #[cfg(test)]
+    mod check_semver_version {
+        use super::*;
+
+        #[test]
+        fn test_exact_version_match() {
+            assert!(check_semver_version("1.0.0", "1.0.0"));
+            assert!(!check_semver_version("1.0.0", "1.1.0"));
+            assert!(!check_semver_version("1.0.0", "1.0.1"));
+            assert!(!check_semver_version("1.0.0", "2.1.0"));
+            assert!(!check_semver_version("1.0.0", "0.1.0"));
+            assert!(!check_semver_version("1.0.0", "0.0.1"));
+        }
+
+        #[test]
+        fn test_version_mismatch() {
+            assert!(!check_semver_version("2.0.0", "1.0.0"));
+        }
+
+        #[test]
+        fn test_patch_version_greater() {
+            assert!(check_semver_version("^1.0.0", "1.0.1"));
+        }
+
+        #[test]
+        fn test_minor_version_greater() {
+            assert!(check_semver_version("^1.0.0", "1.1.0"));
+        }
+
+        #[test]
+        fn test_major_version_less() {
+            assert!(!check_semver_version("1.0.0", "0.9.0"));
+        }
+
+        #[test]
+        fn test_prerelease_less_than_release() {
+            assert!(!check_semver_version("1.0.0", "1.0.0-alpha"));
+        }
+
+        #[test]
+        fn test_prerelease_version_specific_match() {
+            assert!(check_semver_version("1.0.0-alpha", "1.0.0-alpha"));
+        }
+
+        #[test]
+        fn test_build_metadata_ignored() {
+            assert!(check_semver_version("1.0.0", "1.0.0+build.1"));
+        }
+
+        #[test]
+        fn test_greater_than_requirement() {
+            assert!(check_semver_version(">1.0.0", "1.1.0"));
+        }
+
+        #[test]
+        fn test_less_than_requirement_fails_when_greater() {
+            assert!(!check_semver_version("<1.0.0", "1.1.0"));
+        }
+
+        #[test]
+        fn test_caret_requirement_matches_minor_update() {
+            assert!(check_semver_version("^1.1.0", "1.2.0"));
+        }
+
+        #[test]
+        fn test_tilde_requirement_matches_patch_update() {
+            assert!(check_semver_version("~1.0.0", "1.0.1"));
+        }
+
+        #[test]
+        fn test_range_requirement_inclusive() {
+            assert!(check_semver_version(">=1.0.0, <2.0.0", "1.5.0"));
+        }
+
+        #[test]
+        fn test_pre_release_specific_match() {
+            assert!(check_semver_version("1.0.0-alpha.1", "1.0.0-alpha.1"));
+        }
+
+        #[test]
+        fn test_pre_release_non_match_when_requiring_release() {
+            assert!(!check_semver_version("1.0.0", "1.0.0-alpha.1"));
+        }
+
+        // That's not our choice. `semver` does not support `||` operator.
+        // Only asserting here to ensure this behavior (which match our docs).
```

Comment:
If we end up choosing to support the `||` operator:

```suggestion
        // TODO(ologbonowiwi): we need to handle manually the `||` operator.
        // We can create our own `semver` internal crate, which
        // relies on `semver` but has specific things, such as adding
        // the `||` operator, and using `=` as the default instead of `^`.
```

Reply:
Maybe this can be done on an additional PR to release a more mature version of `required_version`, even if it is still unstable. But idk if we need to support `||` or we're good with `semver` default behavior.